docs/man/fdupdate.1
.\" Process this file with
.\" groff -man -Tascii fdmanager.8
.\"
.TH FDUPDATE 8 "MARCH 2002" FramerD "FramerD Documentation"
.SH NAME
fdupdate \- installs files
.SH SYNOPSIS
.B fdupdate 
.I dest source 

.B fdupdate --overwrite
.I dest source 

.B fdupdate 
.I destdir 
.I sources
.B ...

.B fdupdate --overwrite
.I dest source 

.B fdupdate --overwrite
.I destdir 
.I sources
.B ...

.SH DESCRIPTION
.B fdupdate
copies the file
.I source
to the file or directory
.I dest
or the files
.I sources
to the destination directory
.I destdir
being clever about writing through links and retaining
group, mode, date, group, and owner (if possible) of the target.
Unless 
.B --overwrite
is specified, this will not overwrite a destination file
which is newer than its source.  If the file being updated is new,
it uses the environment variables
.B FRAMERD_GROUP
and
.B FRAMERD_USER
to specify the user and group (if possible).

.SH OPTIONS
.IP "--overwrite"
overwrites the destination file even if it is newer than the source.

.SH FILES
No files
.SH BUGS
None known.
.SH Revision Info
.B $Id: fdupdate.1,v 1.3 2003/11/14 14:49:09 haase Exp $
.SH AUTHOR
Ken Haase < kh@beingmeta.com >




NEWS
New in FramerD 2.6

* Numerous bug fixes and expanded error handling
* Added new kind of file index which scales better to multi-million
  key file indices and is based on multiplicative hashing
* General improvements to FDSCRIPT module system
* Defined more core primitive inference procedures especially around
  structured (assoc/pair) slots
* Added generic facility for file/module reloading
* Various frame operation  optimizations
* Added new primitives for accessing serial ports (including USB serial ports)
* Numerous fixes and additions to arithmetic functions in FDScript
* Improvements to index caching performance
* Added compound indices which combine multiple indices with a shared cache.
* Added subservers on index servers, allowing a server to provide
  several distinct index tables
* Provided C API access to ints and strings in config files
* New, simpler command line parsing API
* Use long longs as OIDs if possible
* Added support for largefiles (> 2GB) for pools and indices
* Added synchronization for pools and indices
* Use file memory mapping for file pools and indices for dramatically
  improved performance in many cases;
* Further optimiziations to choice/set handling
* Added "schema slotmaps" to reduce memory footprint for homogenous databases
* Extensions to support extensive malloc debugging
* Introduced interned ZSTRINGS as a core data type
* Extended the frame data model by adding *adjuncts* which allow
  index mappings to specify particular slots of for particular pools.
* Separated out get-methods and compute-methods
* Different pools and indices on the same server now get their own
  network connection
* Added finer grained locks for pool and index access to improve SMP
  performance.
* Optimized index retrieval to return sorted choices
* Added tools to manage malformed UTF-8 text (important for DMOZ integration)
* Improved ability to preload indices to improve performance
* Fixes to HTML/XML integration with special attention to XML
* Fix mime parsing with BASE64 encoding
* Added tools for parsing structured (tab, space, etc) plaintext
* Separated out the FDTEXT and FDWWW libraries
* Made FDXML actually be usable and added callout models to have FDXML
  call Scheme to generate content.


New in FramerD 2.4

This is the first stable release of FramerD since development moved
from MIT to beingmeta.  It includes numerous bug fixes and exhaustive
memory testing (with Parasoft's Insure++ product).

Extensive work was done to find bugs in multithreaded behavior and we
(at beingmeta) now routinely take advantage of SMP hardware to improve
throughput of FDScript programs.  There have also been substantial
performance improvements in both set/choice operations and in the
provision of primitives for prefetching objects and sets of index
entries.

This release also provides additional documentation including a full
set of man pages.

Two major new components of this release are fdmanager, a program for
managing multiple fdservers and mod_fdserv, an Apache module for
maintaining persistent multi-threaded servlets, written in FDscript or
FDXML, for responding to document requests.

New in FramerD 2.2

* substantially improved memory performance
* optimized set/choice operations permitting log or linear time
  operations in most cases
* robust and efficient utilization of multiple processors on SMP
  (symmetric multi processing) architectures
* better integration with Apache for dynamic generation
  of web pages with the fdcgi and fdxml executables
* improved mime parsing with handling of Unicode data in
  headers
* plugin architectures for adding new kinds of pools and indices
* extended subjob and networking socket support (subjob support
  remains incomplete under WIN32)
* builtin MIME, XML, and HTML parsers for dealing with external
  textual data
* extension of character operations to full Unicode and provisions for
  interpreting compound Unicode characters

New in FramerD 2.1

New memory allocation model which both saves space and time and
allows for per-thread free lists, speeding up multi-threaded programming.

Numerous changes to make multi-threaded and SMP programming more robust.

Optimization of choice types for space and time, allowing especially
compact representation of homogenous choices and especially fast set
operations when homogeous types are atomic (e.g. fixnums, symbols,
OIDs).

New indexing model allows the specification of `expanders' for
particular slots.  This is a more general way of customizing indexing
and searching behaviour.

New text primitive MORPHRULE for defining more general morphological
processors than SUFFIXRULE.

New implementation of SLOTMAP datatype allows for shared frame
schemas, to permit more compact object representations.

New, simpler, XML/HTML parser.  Support for unparsing generic XML back
into text documents.  New executable FDXML generates HTML or XML from
a source document (in XML) with programmatic inclusions.

Apache module mod_fdcgi (based on mod_fastcgi) provides for persistent
CGI servers implicitly created based on fdcgi scripts or fdxml documents.

New in FramerD 2.0

Internal Unicode support is much more pervasive and robust and a large
number of external character encodings is supported, in addition to
the ability to read Unicode Consortium "mapping" files to get more.
In addition, the character properties (upper, lower, digit, etc) have
now been extended to all of the Unicode BMP (Basic Multilingual
Plane).

A new init/configuration file interface makes it much easier to
customize FramerD at your site or for your personal use.  Config files
can now specify additions to (rather than just assignments for) global
variables.  You can either load config files by hand using
LOAD-CONFIG, or specify them on the command line with --config=file.

FramerD database (pool) servers can now serve multiple pools at the
same time.  Unlike the facility in FramerD 0.9, these pools do not

src/exe/fdupdate.c
#include <framerd/dtypes.h>
#include <stdio.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <utime.h>

static char usage[]=
  "fdupdate <dest> <source> [source]..]\nfdupdate --overwrite <dest> <source> [source]...\n";

static void install_file(char *from,char *to,int update)
{
  FILE *in=fd_fopen_locked(from,"r+b",0), *out;
  int  bufsize=65536*4, ret_value=0, bytes=0, new_file=0, through_link=0;
  char *buf=fd_xmalloc(bufsize), *realname=NULL;
  struct stat from_info, to_info;
  if (fd_symbolic_linkp(to)) {
    char *realsrc=fd_get_real_pathname(from);
    realname=fd_get_real_pathname(to);
    if (strcmp(realsrc,realname) == 0) {
      fd_warn(_("Not copying since %s is already linked to %s"),to,from);
      fd_xfree(realsrc); fd_xfree(realname);
      return;}
    through_link=1;}
  else realname=fd_strdup(to);
  stat(from,&from_info); 
  new_file=(fd_file_existsp(realname) == 0);
  if ((update) && (new_file==0)) {
    stat(realname,&to_info);
    if (to_info.st_mtime >= from_info.st_mtime) {
      fd_warn(_("Not overwriting newer '%s'"),realname);
      return;}}
  if (through_link)
    fd_warn(_("Copying %s to %s (through link %s)"),from,realname,to);
  else fd_warn(_("Copying %s to %s"),from,realname);
  out=fopen(realname,"wb");
  if (errno) {perror("Start of binary copy"); FD_CLEAR_ERR();}
  if (in == NULL) 
    fd_raise_detailed_exception(fd_FileOpenFailed,from);
  else if (out == NULL) 
    fd_raise_detailed_exception(fd_FileOpenWFailed,to);
  else while ((ret_value=fread(buf,sizeof(char),bufsize,in)) ||
	      (!(feof(in)))) {
    bytes=bytes+ret_value; fwrite(buf,sizeof(char),ret_value,out);}
  fclose(out); fclose(in); fd_xfree(buf);
  if (new_file) {
    char *gname, *uname; gid_t gid; uid_t uid;
    gname=getenv("FRAMERD_GROUP"); if (gname == NULL) gname="framerd";
    gid=fd_get_gid(gname);
    uname=getenv("FRAMERD_USER"); if (uname == NULL) uname="framerd";
    uid=fd_get_uid(uname);
    if (gid < 0) gid=from_info.st_gid;
    chown(realname,uid,gid);}

  {struct utimbuf wbuf;
   wbuf.actime=from_info.st_atime; wbuf.modtime=from_info.st_mtime;
   if (utime(realname,&wbuf) < 0) perror("utime");}
  fd_xfree(realname);
}

static int ends_in_slashp(char *string)
{
  int len=strlen(string);
  if (len == 0) return 0;
  else if (string[len-1] == '/') return 1;
  else return 0;
}

int main(int argc,char *argv[])
{
  int update=1; char *target; int source_start;
  if (argc < 2) {fputs(usage,stderr); exit(1);}
  if (argv[1][0] == '-') {
    if (strcmp(argv[1],"--overwrite") == 0) update=0;
    target=argv[2]; source_start=3;}
  else {
    target=argv[1]; source_start=2;}
  fd_initialize_dtypes();
  if (argc <= source_start) {fputs(usage,stderr); exit(1);}
  else if ((argc-source_start == 1) && (!(fd_directoryp(target))))
    install_file(argv[source_start],target,update);
  else {
    int need_slashp=(ends_in_slashp(target) == 0), dirlen=strlen(target);
    int i=source_start; while (i < argc) {
      char *base=fd_basename(argv[i],1);
      char *dest_path=fd_xmalloc(strlen(base)+dirlen+2);
      strcpy(dest_path,target);
      if (need_slashp) strcat(dest_path,"/");
      strcat(dest_path,base);
      install_file(argv[i],dest_path,update);
      fd_xfree(dest_path); i++;}}
  return 0;
}
